File: typescript/src/client/refresh.ts
/**
 * Token refresh.
 *
 * Refresh is either PROACTIVE (lifetime < threshold, default 25%) or
 * FORCED (`now > force_online_after`). The two paths behave differently
 * on network failure:
 *
 *   - Proactive: swallow the error. We still have a perfectly valid
 *     token; coming back later is fine. The caller may surface a log
 *     line but MUST NOT fail the user.
 *   - Forced: either succeed, enter grace (if not already in grace and
 *     a grace window is configured), or surface `RequiresOnlineRefresh`.
 *     Grace is a mercy window for a server being temporarily unreachable
 *     past the hard-deadline the issuer chose.
 *
 * A successful refresh always clears `graceStartSec` — whatever outage
 * caused grace is over, the issuer and client are back in sync.
 */

import { clientErrors, LicensingClientError } from './errors.ts';
import { EMPTY_STATE, type TokenStore } from './token-store.ts';
import { postJson, type TransportOptions } from './transport.ts';
import { peek } from './validate.ts';

export interface RefreshOptions extends TransportOptions {
  readonly store: TokenStore;
  /** Current wall-clock unix seconds. Injected for testability. */
  readonly nowSec: number;
  /** Lifetime fraction below which a proactive refresh fires. Default
   *  0.25. Set to 0 to disable proactive behavior; set to 1.0 to refresh
   *  every call. */
  readonly proactiveThreshold?: number;
  /** Grace-on-unreachable window in seconds. Default 7 days (604_800).
   *  Set to 0 to disable grace entirely — forced refresh failures then
   *  surface `IssuerUnreachable` directly. */
  readonly graceWindowSec?: number;
  /** Path under `baseUrl`. Default `/api/licensing/v1/refresh`. */
  readonly path?: string;
}

export type RefreshOutcome =
  | { readonly kind: 'refreshed'; readonly token: string }
  | { readonly kind: 'not-due'; readonly token: string }
  | { readonly kind: 'grace-entered'; readonly token: string; readonly graceStartSec: number }
  | { readonly kind: 'grace-continued'; readonly token: string; readonly graceStartSec: number };

/** Shape of the issuer's success `data`. Only `token` is mandatory; other
 *  fields (usage_id, license_id) may be returned but aren't needed here. */
interface RefreshResponseData {
  readonly token: unknown;
}

/**
 * Attempt to refresh the stored token. Returns a tagged outcome describing
 * what happened; consumers that only want a boolean "was a network call
 * made" can look at `outcome.kind`.
 *
 * Throws {@link LicensingClientError} for unrecoverable states:
 *   - `NoToken`: nothing stored; activate first.
 *   - `GraceExpired`: grace window elapsed without a successful contact.
 *   - `RequiresOnlineRefresh`: forced deadline past AND grace disabled
 *     AND network failed — there's no legitimate way forward but to
 *     surface the hard error.
 *   - Any issuer-returned error (LicenseRevoked, SeatLimitExceeded, ...).
 */
export async function refresh(opts: RefreshOptions): Promise<RefreshOutcome> {
  const state = await opts.store.read();
  const currentToken = state.token;
  if (currentToken === null) {
    throw clientErrors.noToken();
  }

  // Peek is cheap and tells us if a refresh is due. We deliberately use
  // the unverified peek: if the token is tampered, the next `validate()`
  // will catch it — `refresh()` operates on the token the store says we
  // have. Using `validate()` here would also require the caller to pass
  // a registry/keys, which couples the refresh API to the crypto surface.
  const metadata = peek(currentToken);
  const graceWindow = opts.graceWindowSec ?? 7 * 24 * 3600;
  const threshold = opts.proactiveThreshold ?? 0.25;

  // First: check grace already-active. If grace is set and expired, fail
  // closed. If it's set and still valid but we're past force_online_after,
  // we still try the network — a successful refresh clears grace.
  if (state.graceStartSec !== null) {
    if (graceWindow > 0 && opts.nowSec > state.graceStartSec + graceWindow) {
      throw clientErrors.graceExpired(
        `grace started at ${state.graceStartSec}, window ${graceWindow}s, now ${opts.nowSec}`,
      );
    }
  }

  const forced = metadata.forceOnlineAfter !== null && metadata.forceOnlineAfter <= opts.nowSec;
  const proactive = shouldProactiveRefresh(metadata, opts.nowSec, threshold);

  if (!forced && !proactive && state.graceStartSec === null) {
    return { kind: 'not-due', token: currentToken };
  }

  // Attempt the network call.
  try {
    const data = await postJson<RefreshResponseData>(
      opts.path ?? '/api/licensing/v1/refresh',
      { token: currentToken },
      opts,
    );
    if (typeof data.token !== 'string' || data.token.length === 0) {
      throw clientErrors.invalidTokenFormat(
        'issuer returned success but no refreshed token string',
      );
    }
    // Success — clear any grace state.
    await opts.store.write({ token: data.token, graceStartSec: null });
    return { kind: 'refreshed', token: data.token };
  } catch (err) {
    // Only IssuerUnreachable is a candidate for grace. Every other
    // issuer error (revoked, seat-limit, ...) is an authoritative
    // response we must NOT mask with grace — the license actually is
    // in that state.
    if (!(err instanceof LicensingClientError) || err.code !== 'IssuerUnreachable') {
      throw err;
    }
    return await handleUnreachable(
      currentToken,
      state.graceStartSec,
      opts,
      forced,
      graceWindow,
      err,
    );
  }
}

/** Extract-and-named for readability. Decides grace behavior when the
 *  network is down during a refresh attempt. `currentToken` is taken as
 *  a non-nullable string — the caller already proved it's not null. */
async function handleUnreachable(
  currentToken: string,
  existingGraceStartSec: number | null,
  opts: RefreshOptions,
  forced: boolean,
  graceWindow: number,
  err: LicensingClientError,
): Promise<RefreshOutcome> {
  if (!forced) {
    // Proactive refresh that failed: swallow. Caller may log.
    return { kind: 'not-due', token: currentToken };
  }
  if (graceWindow === 0) {
    // Caller opted out of grace. Surface the hard error.
    throw clientErrors.requiresOnlineRefresh(
      `forced refresh failed, grace disabled: ${err.message}`,
    );
  }
  // Forced + network down + grace enabled.
  if (existingGraceStartSec !== null) {
    // Already in grace. Double-check window (we checked on entry, but
    // time may have moved). Past window → GraceExpired.
    if (opts.nowSec > existingGraceStartSec + graceWindow) {
      throw clientErrors.graceExpired(
        `grace started at ${existingGraceStartSec}, window ${graceWindow}s, now ${opts.nowSec}`,
      );
    }
    return {
      kind: 'grace-continued',
      token: currentToken,
      graceStartSec: existingGraceStartSec,
    };
  }
  // Enter grace.
  const graceStartSec = opts.nowSec;
  await opts.store.write({ token: currentToken, graceStartSec });
  return { kind: 'grace-entered', token: currentToken, graceStartSec };
}

/** Return true when the proactive-refresh condition holds. Lifetime is
 *  computed from (nbf → exp) rather than (iat → exp) because nbf is the
 *  effective start: a token issued at T1 but nbf=T2 becomes usable at T2,
 *  and "25% remaining" should be relative to the usable window. */
function shouldProactiveRefresh(
  metadata: { readonly nbf: number; readonly exp: number },
  nowSec: number,
  threshold: number,
): boolean {
  if (threshold <= 0) return false;
  const lifetime = metadata.exp - metadata.nbf;
  if (lifetime <= 0) return true; // degenerate — refresh immediately
  const remaining = metadata.exp - nowSec;
  if (remaining <= 0) return true; // already expired, caller's validate() will fail anyway
  return remaining / lifetime < threshold;
}

// Re-export EMPTY_STATE so callers constructing the module graph don't
// need a separate import just for the token-store types.
export { EMPTY_STATE };

File: licensing/client/refresh.go
package client

import (
	"errors"
	"fmt"
)

// RefreshOptions configures the refresh call.
type RefreshOptions struct {
	Store              TokenStore
	Path               string
	Transport          TransportOptions
	NowSec             int64
	ProactiveThreshold float64
	GraceWindowSec     int64
	GraceWindowSet     bool
}

func (o RefreshOptions) path() string {
	if o.Path != "" {
		return o.Path
	}
	return "/api/licensing/v1/refresh"
}

func (o RefreshOptions) threshold() float64 {
	if o.ProactiveThreshold > 0 {
		return o.ProactiveThreshold
	}
	return 0.25
}

func (o RefreshOptions) graceWindow() int64 {
	if o.GraceWindowSet {
		return o.GraceWindowSec
	}
	return 604800
}

// RefreshKind is the tagged-union discriminator for RefreshOutcome.
type RefreshKind string

// RefreshKind values. `Refreshed` is the happy path; `NotDue` means the
// token is still fresh and no network call was made; the `Grace*` variants
// report that the issuer was unreachable but the client is still within
// the configured grace window.
const (
	RefreshKindRefreshed      RefreshKind = "refreshed"
	RefreshKindNotDue         RefreshKind = "not-due"
	RefreshKindGraceEntered   RefreshKind = "grace-entered"
	RefreshKindGraceContinued RefreshKind = "grace-continued"
)

// RefreshOutcome is the return type of Refresh.
type RefreshOutcome struct {
	GraceStartSec *int64
	Kind          RefreshKind
	Token         string
}

type refreshRequest struct {
	Token string `json:"token"`
}

type refreshResponse struct {
	Token string `json:"token"`
}

// Refresh proactively refreshes or force-refreshes the stored token.
// Returns a tagged outcome describing what happened.
func Refresh(opts RefreshOptions) (*RefreshOutcome, error) {
	state, err := opts.Store.Read()
	if err != nil {
		return nil, fmt.Errorf("read token store: %w", err)
	}
	if state.Token == "" {
		return nil, NoToken("")
	}

	peek, err := Peek(state.Token)
	if err != nil {
		return nil, err
	}

	graceWindow := opts.graceWindow()

	// If in grace and grace has expired, fail immediately.
	if state.GraceStartSec != nil {
		if graceWindow > 0 && opts.NowSec >= *state.GraceStartSec+graceWindow {
			return nil, GraceExpired("")
		}
	}

	forced := peek.ForceOnlineAfter != nil && opts.NowSec >= *peek.ForceOnlineAfter
	proactive := shouldProactiveRefresh(peek, opts.NowSec, opts.threshold())
	inGrace := state.GraceStartSec != nil

	if !forced && !proactive && !inGrace {
		return &RefreshOutcome{Kind: RefreshKindNotDue, Token: state.Token}, nil
	}

	// Attempt online refresh.
	resp, err := PostJSON[refreshResponse](opts.path(), refreshRequest{
		Token: state.Token,
	}, opts.Transport)

	if err != nil {
		// Only IssuerUnreachable triggers grace logic.
		var ce *ClientError
		if !errors.As(err, &ce) || ce.Code != CodeIssuerUnreachable {
			return nil, err
		}

		// Network failure on forced refresh → enter/continue grace.
		if forced {
			if graceWindow == 0 {
				return nil, RequiresOnlineRefresh("")
			}
			if inGrace {
				return &RefreshOutcome{
					Kind:          RefreshKindGraceContinued,
					Token:         state.Token,
					GraceStartSec: state.GraceStartSec,
				}, nil
			}
			// Enter grace.
			gs := opts.NowSec
			if writeErr := opts.Store.Write(StoredTokenState{
				Token:         state.Token,
				GraceStartSec: &gs,
			}); writeErr != nil {
				return nil, fmt.Errorf("write grace state: %w", writeErr)
			}
			return &RefreshOutcome{
				Kind:          RefreshKindGraceEntered,
				Token:         state.Token,
				GraceStartSec: &gs,
			}, nil
		}

		// Proactive refresh failure → swallow.
		return &RefreshOutcome{Kind: RefreshKindNotDue, Token: state.Token}, nil
	}

	// Success — clear grace.
	if err := opts.Store.Write(StoredTokenState{Token: resp.Token}); err != nil {
		return nil, fmt.Errorf("write refreshed token: %w", err)
	}
	return &RefreshOutcome{Kind: RefreshKindRefreshed, Token: resp.Token}, nil
}

func shouldProactiveRefresh(peek *PeekResult, nowSec int64, threshold float64) bool {
	lifetime := peek.Exp - peek.Nbf
	if lifetime <= 0 {
		return true
	}
	remaining := peek.Exp - nowSec
	if remaining <= 0 {
		return true
	}
	return float64(remaining)/float64(lifetime) < threshold
}
